### PR TITLE
ARROW-15126: [C++] Support Null type as group keys

### PR DESCRIPTION
**Pull request by Crystrix**

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/hash_aggregate.cc [hunk 1]
```diff
@@ -147,10 +152,13 @@ struct GrouperImpl : Grouper {
       if (it_success.second) {
         // new key; update offsets and key_bytes
         ++num_groups_;
-        auto next_key_offset = static_cast<int32_t>(key_bytes_.size());
-        key_bytes_.resize(next_key_offset + key_length);
-        offsets_.push_back(next_key_offset + key_length);
-        memcpy(key_bytes_.data() + next_key_offset, key.c_str(), key_length);
+        if (key_length != 0) {
```
Comment (edponce) [inline cpp/src/arrow/compute/kernels/hash_aggregate.cc:155]: Nit: `key_length > 0` is a bit more representable given that `key_length` is a signed int.

File: cpp/src/arrow/compute/kernels/row_encoder.h [hunk 2]
```diff
@@ -231,6 +231,26 @@ struct VarLengthKeyEncoder : KeyEncoder {
   std::shared_ptr<DataType> type_;
 };
 
+struct NullKeyEncoder : KeyEncoder {
+  void AddLength(const Datum&, int64_t batch_length, int32_t* lengths) override {
+    return;
+  }
```
Comment (edponce) [inline cpp/src/arrow/compute/kernels/row_encoder.h:237]: No need to `return` as it is a void method, empty braces will suffice.

Comment (edponce) [reply cpp/src/arrow/compute/kernels/row_encoder.h:237]: Similarly, for `AddLengthNull` and `EncodeNull`.

**Review by edponce — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/hash_aggregate.cc [hunk 3]
```diff
@@ -147,10 +152,13 @@ struct GrouperImpl : Grouper {
...
+          // Skip if there are no keys
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/hash_aggregate.cc:156]: I'm not sure I understand this comment, do you need to move it up just above the `if`?

File: cpp/src/arrow/compute/kernels/hash_aggregate_test.cc [hunk 4]
```diff
@@ -614,6 +616,11 @@ TEST(Grouper, RandomStringInt64DoubleInt32Keys) {
   }
 }
 
+TEST(Grouper, NullKeys) {
+  TestGrouper g({null()});
+  g.ExpectConsume("[[null], [null]]", "[0, 0]");
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/hash_aggregate_test.cc:621]: Is it possible to add a test with several keys and where one of them has the null type?

File: cpp/src/arrow/compute/kernels/hash_aggregate.cc [hunk 5]
```diff
@@ -197,7 +205,7 @@ struct GrouperFastImpl : Grouper {
 #if ARROW_LITTLE_ENDIAN
     for (size_t i = 0; i < keys.size(); ++i) {
       const auto& key = keys[i].type;
-      if (is_large_binary_like(key->id())) {
+      if (is_large_binary_like(key->id()) || key->id() == Type::NA) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/hash_aggregate.cc:208]: I'm not sure I understand, why cannot null keys be used here?

File: cpp/src/arrow/compute/kernels/hash_aggregate.cc (hunk 1 above)
Comment (Crystrix) [reply cpp/src/arrow/compute/kernels/hash_aggregate.cc]: Fixed

File: cpp/src/arrow/compute/kernels/row_encoder.h (hunk 2 above)
Comment (Crystrix) [reply cpp/src/arrow/compute/kernels/row_encoder.h]: Fixed

File: cpp/src/arrow/compute/kernels/hash_aggregate.cc (hunk 3 above)
Comment (Crystrix) [reply cpp/src/arrow/compute/kernels/hash_aggregate.cc]: The comment is moved up.

File: cpp/src/arrow/compute/kernels/hash_aggregate_test.cc (hunk 4 above)
Comment (Crystrix) [reply cpp/src/arrow/compute/kernels/hash_aggregate_test.cc:620]: More test cases including this one are added.

File: cpp/src/arrow/compute/kernels/hash_aggregate.cc (hunk 5 above)
Comment (Crystrix) [reply cpp/src/arrow/compute/kernels/hash_aggregate.cc]: In the previous commits, GrouperFastImpl for null type is not implemented yet, so null should be skipped for GrouperFastImpl. Now the null type is added to GrouperFastImpl, so the condition is removed.

File: cpp/src/arrow/compute/exec/key_encode.h [hunk 6]
```diff
@@ -53,13 +53,19 @@ class KeyEncoder {
...
+    KeyColumnMetadata(bool is_fixed_length_in, uint32_t fixed_length_in,
+                      bool is_null_type_in = false)
+        : is_fixed_length(is_fixed_length_in),
+          is_null_type(is_null_type_in),
+          fixed_length(fixed_length_in) {}
     /// Is column storing a varying-length binary, using offsets array
     /// to find a beginning of a value, or is it a fixed-length binary.
     bool is_fixed_length;
+    /// Is column null type
+    bool is_null_type;
```
Comment (Crystrix) [inline cpp/src/arrow/compute/exec/key_encode.h:65]: A flag of null type is added. As the `fixed_length` of a null type column is zero like a bit vector which leads to confusion.

**Comment by Crystrix** [comment]
@pitrou @edponce, Could you please help to review?
The main changes are updating based on comments and supporting the null type key in the `GrouperFastImpl`.

File: cpp/src/arrow/compute/kernels/hash_aggregate.cc [hunk 7]
```diff
@@ -322,14 +333,20 @@ struct GrouperFastImpl : Grouper {
...
+      const uint8_t* non_nulls = NULLPTR;
+      const uint8_t* fixedlen = NULLPTR;
+      const uint8_t* varlen = NULLPTR;
+
+      // Skip if the key's type is NULL
+      if (key_types_[icol]->id() != Type::NA) {
+        if (batch[icol].array()->buffers[0] != NULLPTR) {
+          non_nulls = batch[icol].array()->buffers[0]->data();
+        }
+        auto array = batch[icol].array();
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/hash_aggregate.cc:345]: What is this variable for? Did you mean to use it?

Comment (pitrou) [inline cpp/src/arrow/compute/kernels/hash_aggregate.cc:345]: What is this variable for? Did you mean to use it?

File: cpp/src/arrow/compute/exec/key_compare.cc [hunk 8]
```diff
@@ -334,6 +334,14 @@ void KeyCompare::CompareColumnsToRows(uint32_t num_rows_to_compare,
   bool is_first_column = true;
   for (size_t icol = 0; icol < cols.size(); ++icol) {
     const KeyEncoder::KeyColumnArray& col = cols[icol];
+    if (col.metadata().is_null_type) {
+      // If this null type col is the first column, the match_bytevector_A needs to be
+      // initialized with 0xFF. Otherwise, the calculation can be skipped
```
Comment (pitrou) [inline cpp/src/arrow/compute/exec/key_compare.cc:339]: Why 0xFF? Can you explain?

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/hash_aggregate.cc (hunk 7 above)
Comment (Crystrix) [reply cpp/src/arrow/compute/kernels/hash_aggregate.cc]: Removed

Comment (Crystrix) [reply cpp/src/arrow/compute/kernels/hash_aggregate.cc]: Removed

File: cpp/src/arrow/compute/exec/key_hash.cc [hunk 9]
```diff
@@ -286,34 +286,40 @@ void Hashing::HashMultiColumn(const std::vector<KeyEncoder::KeyColumnArray>& col
...
-                                      byte_temp_buf.mutable_data(),
-                                      cols[icol].bit_offset(1));
-      }
-      Hashing::hash_fixed(
-          ctx->hardware_flags, num_rows, col_width == 0 ? 1 : col_width,
-          col_width == 0 ? byte_temp_buf.mutable_data() : cols[icol].data(1),
-          is_first ? out_hash : hash_temp_buf.mutable_data());
+    // Set the hash value as zero for a null type col
+    if (cols[icol].metadata().is_null_type) {
+      uint32_t* dst_hash = is_first ? out_hash : hash_temp_buf.mutable_data();
```
[the remark this answers is not in the document]
Comment (Crystrix) [reply cpp/src/arrow/compute/exec/key_hash.cc]: The calculation of `dst_hash` is moved to the start of each iteration and used in the following steps if my understanding is correct.

File: cpp/src/arrow/compute/kernels/hash_aggregate.cc [hunk 10]
```diff
@@ -237,6 +245,9 @@ struct GrouperFastImpl : Grouper {
       } else if (is_binary_like(key->id())) {
         impl->col_metadata_[icol] =
             arrow::compute::KeyEncoder::KeyColumnMetadata(false, sizeof(uint32_t));
+      } else if (key->id() == Type::NA) {
+        impl->col_metadata_[icol] =
+            arrow::compute::KeyEncoder::KeyColumnMetadata(true, 0, true);
```
[the remark this answers is not in the document]
Comment (Crystrix) [reply cpp/src/arrow/compute/kernels/hash_aggregate.cc]: Added

**Review by pitrou — APPROVED** [review]
LGTM. Thanks for the update @Crystrix.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
@Crystrix sorry, it looks like merging your other PR means this one needs rebasing now

**Comment by Crystrix** [comment]
> @Crystrix sorry, it looks like merging your other PR means this one needs rebasing now

@lidavidm, the rebasing is done.

**Comment by lidavidm** [comment]
@github-actions autotune

**Comment by lidavidm** [comment]
@github-actions rebase

**Comment by lidavidm** [comment]
Let's hope that actually kicks off the tests…sorry for the trouble here.

**Comment by lidavidm** [comment]
Hmm, nope. Sorry @Crystrix, do you mind trying to rebase/amend a commit to see if it'll trigger the builds for you? I wonder if CI ignores 'bot' pushes.

**Comment by Crystrix** [comment]
@lidavidm, new builds are triggered, the failure should be irrelevant.

**Comment by lidavidm** [comment]
Thanks!

**Pull request closed without merging**